---
title: "Cell type annotation"
output: html_notebook
---

```{r, include=FALSE}
library(scales)
library(cowplot)
library(RCurl)
library(SingleR)
# Load libraries
library(scater)
library(Seurat)
library(tidyverse)
library(cowplot)
library(Matrix.utils)
library(edgeR)
library(dplyr)
library(magrittr)
library(Matrix)
library(purrr)
library(reshape2)
library(S4Vectors)
library(tibble)
library(SingleCellExperiment)
library(pheatmap)
library(apeglm)
library(png)
library(DESeq2)
library(RColorBrewer)
```


# We start with the filtered dataset from quality control file

```{r}
load(file="data/seurat_filtered.RData")
```
```{r}
filtered_seurat
```
# We get the counts and meta data from seurat object

```{r}
counts <- filtered_seurat@assays$RNA@counts 

metadata <- filtered_seurat@meta.data
```



```{r}
sce <- as.SingleCellExperiment(DietSeurat(filtered_seurat))
```

```{r}
monaco.ref <- celldex::MonacoImmuneData()
```

```{r}
monaco.main <- SingleR(test = sce,assay.type.test = 1,ref = monaco.ref,labels = monaco.ref$label.main)
```

```{r}
table(monaco.main$pruned.labels)
```
```{r}
metadata$cluster_id <- factor(monaco.main$labels)
```
```{r}
metadata
```
```{r}
sce
```
```{r}
sce@metadata <- list(t=metadata)
```
```{r}
sce@metadata$t
```

```{r}
colData(sce)$cluster_id <- sce@metadata$t$cluster_id
colData(sce)$sample_id <- factor(paste0(colData(sce)$orig.ident, colData(sce)$nUMI))
```

```{r}
dim(colData(sce))

head(colData(sce))
```

```{r}
# Named vector of cluster names
kids <- purrr::set_names(levels(sce$cluster_id))
kids
```
```{r}
# Total number of clusters
nk <- length(kids)
nk

# Named vector of sample names
sids <- purrr::set_names(levels(sce$sample_id))

# Total number of samples 
ns <- length(sids)
ns
```
To perform sample-level differential expression analysis, we need to generate sample-level metadata. To do this, we will reorder samples in the single-cell metadata to match the order of the factor levels of the sample ID, then extract only the sample-level information from the first cell corresponding to that sample.

```{r}
# Generate sample level metadata

## Determine the number of cells per sample
# table(sce$sample_id)

## Turn named vector into a numeric vector of number of cells per sample
n_cells <- as.numeric(table(sce$sample_id))

## Determine how to reoder the samples (rows) of the metadata to match the order of sample names in sids vector
m <- match(sids, sce$sample_id)

## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
ei <- data.frame(colData(sce)[m, ], 
                  n_cells, row.names = NULL) %>% 
                select(c("orig.ident", "nUMI", "sample_id","n_cells"))
ei


```
Since we did quality control at very begining, we don't need to do it here. However, if you have not, we should do QC here.

```{r}
# Aggregate the counts per sample_id and cluster_id

# Subset metadata to only include the cluster and sample IDs to aggregate across
groups <- colData(sce)[, c("cluster_id", "sample_id")]

# Aggregate across cluster-sample groups
pb <- aggregate.Matrix(t(counts(sce)), 
                       groupings = groups, fun = "sum") 

class(pb)

dim(pb)

pb[1:6, 1:6]
```
The output of this aggregation is a sparse matrix, and when we take a quick look, we can see that it is a gene by cell type-sample matrix.


To perform DE analysis on a per cell type basis, we need to wrangle our data in a couple ways. We need to do the following steps:

  + Split our data by cell type
  + Transform the matrix so that the genes are the row names and the samples are the column names

We will split our data by cell type; however, not always do all samples contain cells of every cell type. To determine which samples are present for each cell type we can run the following:
```{r}
splitf <- sapply(stringr::str_split(rownames(pb), 
                                    pattern = "_",  
                                    n = 2), 
                 `[`, 1)
```

Now we can turn the matrix into a list that is split into count matrices for each cluster, then transform each data frame so that rows are genes and columns are the samples.

```{r}
# Turn into a list and split the list into components for each cluster and transform, so rows are genes and columns are samples and make rownames as the sample IDs
pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
        lapply(function(u) 
                set_colnames(t(u), 
                             stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+")))

class(pb)

# Explore the different components of list
str(pb)
```
The counts per sample for each cluster can be checked:

```{r}
# Print out the table of cells in each cluster-sample group
options(width = 100)
table(sce$cluster_id, sce$sample_id)[1:6,1:6]
```
Sample-level metadata

To perform the DE analysis, we need metadata for all samples, including cluster ID, sample ID and the condition(s) of interest (group_id), in addition to any other sample-level metadata (e.g. batch, sex, age, etc.). The ei data frame holds the sample ID and condition information, but we need to combine this information with the cluster IDs.

First, we will create a vector of sample names combined for each of the cell type clusters.

```{r}
# Get sample names for each of the cell type clusters

# prep. data.frame for plotting
get_sample_ids <- function(x){
        pb[[x]] %>%
                colnames()
}

de_samples <- map(1:length(kids), get_sample_ids) %>%
        unlist()
```

Then we can get the cluster IDs corresponding to each of the samples in the vector.

```{r}
# Get cluster IDs for each of the samples

samples_list <- map(1:length(kids), get_sample_ids)

get_cluster_ids <- function(x){
        rep(names(pb)[x], 
            each = length(samples_list[[x]]))
}

de_cluster_ids <- map(1:length(kids), get_cluster_ids) %>%
        unlist()
```

Finally, let’s create a data frame with the cluster IDs and the corresponding sample IDs. We will merge together the condition information.

```{r}
# Create a data frame with the sample IDs, cluster IDs and condition

gg_df <- data.frame(cluster_id = de_cluster_ids,
                    sample_id = de_samples)

gg_df <- left_join(gg_df, ei[, c("sample_id", "orig.ident")]) %>% rename("group_id" = orig.ident)


metadata <- gg_df %>%
        dplyr::select(cluster_id, sample_id, group_id) %>% mutate(cluster_id = factor(cluster_id))
        
metadata   
```
Subsetting dataset to cluster(s) of interest

```{r}
# Generate vector of cluster IDs
clusters <- levels(metadata$cluster_id)
clusters
```

We see multiple different immune cell types in our dataset. Let’s perform the DE analysis on B cells, which represent the first element in our vector. Let’s extract the B cells from the vector:

```{r}
clusters[1]
```

We can use this output to run the DE analysis on the B cells. First we can subset the metadata and the counts to only the B cells.

```{r}
# Subset the metadata to only the B cells
cluster_metadata <- metadata[which(metadata$cluster_id == clusters[1]), ]
head(cluster_metadata)

# Assign the rownames of the metadata to be the sample IDs
rownames(cluster_metadata) <- cluster_metadata$sample_id
head(cluster_metadata)

# Subset the counts to only the B cells
counts <- pb[[clusters[1]]]

cluster_counts <- data.frame(counts[, which(colnames(counts) %in% rownames(cluster_metadata))])

# Check that all of the row names of the metadata are the same and in the same order as the column names of the counts in order to use as input to DESeq2
all(rownames(cluster_metadata) == colnames(cluster_counts))        
```

Create DESeq2 object

Now we can create our DESeq2 object to prepare to run the DE analysis. We need to include the counts, metadata, and design formula for our comparison of interest. In the design formula we should also include any other columns in the metadata for which we want to regress out the variation (e.g. batch, sex, age, etc.). We only have our comparison of interest, which is stored as the group_id in our metadata data frame.

```{r}
dds <- DESeqDataSetFromMatrix(cluster_counts, 
                              colData = cluster_metadata, 
                              design = ~ group_id)
```

Run DESeq2


```{r}
# Run DESeq2 differential expression analysis
dds <- DESeq(dds)
```
```{r}
# Plot dispersion estimates
plotDispEsts(dds)
```
Now that we have performed the differential expression analysis, we can explore our results for a particular comparison. To denote our comparison of interest, we need to specify the contrast and perform shrinkage of the log2 fold changes.

Let’s compare the stimulated group relative to the control:

```{r}
resultsNames(dds)
```


```{r}
# Output results of Wald test for contrast for stim vs ctrl
levels(cluster_metadata$group_id)[2]
levels(cluster_metadata$group_id)[1]

contrast <- c("group_id", levels(cluster_metadata$group_id)[2], levels(cluster_metadata$group_id)[1])

# resultsNames(dds)
res <- results(dds, 
               contrast = contrast,
               alpha = 0.05)

res <- lfcShrink(dds, 
                 coef =  2,
                 res=res)
```
```{r}
res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()

# Check results output
res_tbl
```
Table of results for significant genes

Next, we can filter our table for only the significant genes using a p-adjusted threshold of 0.05

```{r}
# Set thresholds
padj_cutoff <- 0.05

# Subset the significant results
sig_res <- dplyr::filter(res_tbl, padj < padj_cutoff) %>%
        dplyr::arrange(padj)

# Check significant genes output
sig_res
```
# Scatter plot  of top 20 genes

```{r}
## ggplot of top genes
normalized_counts <- counts(dds, 
                            normalized = TRUE)

## Order results by padj values
top20_sig_genes <- sig_res %>%
        dplyr::arrange(padj) %>%
        dplyr::pull(gene) %>%
        head(n=20)


top20_sig_norm <- data.frame(normalized_counts) %>%
        rownames_to_column(var = "gene") %>%
        dplyr::filter(gene %in% top20_sig_genes)

gathered_top20_sig <- top20_sig_norm %>%
        gather(colnames(top20_sig_norm)[2:length(colnames(top20_sig_norm))], key = "samplename", value = "normalized_counts")
        
gathered_top20_sig <- inner_join(ei[, c("sample_id", "orig.ident" )], gathered_top20_sig, by = c("sample_id" = "samplename")) %>% rename(group_id = orig.ident)

## plot using ggplot2
ggplot(gathered_top20_sig) +
        geom_point(aes(x = gene, 
                       y = normalized_counts, 
                       color = group_id), 
                   position=position_jitter(w=0.1,h=0), alpha=0.3) +
        scale_y_log10() +
        xlab("Genes") +
        ylab("log10 Normalized Counts") +
        ggtitle("Top 20 Significant DE Genes") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        theme(plot.title = element_text(hjust = 0.5))
```
# Volcano plot

```{r}
## Obtain logical vector where TRUE values denote padj values < 0.05 and fold change > 1.5 in either direction
res_table_thres <- res_tbl %>% 
                  mutate(threshold = padj < 0.05 & abs(log2FoldChange) >= 0.58)
                  
## Volcano plot
ggplot(res_table_thres) +
    geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold)) +
    ggtitle("Volcano plot of stimulated B cells relative to control") +
    xlab("log2 fold change") + 
    ylab("-log10 adjusted p-value") +
    scale_y_continuous(limits = c(0,50)) +
    theme(legend.position = "none",
          plot.title = element_text(size = rel(1.5), hjust = 0.5),
          axis.title = element_text(size = rel(1.25)))                    
```
```{r}
library(org.Hs.eg.db)
```

